src/main/java/com/mparticle/kits/PilgrimSDKKit.java

```diff
@@ -0,0 +1,183 @@
+        messageList.add(new ReportingMessage(this, ReportingMessage.MessageType.APP_STATE_TRANSITION, System.currentTimeMillis(), null));
+        return messageList;
+    }
+
+    @Override
+    public String getName() {
+        return KIT_NAME;
+    }
+
+    @Override
+    public List<ReportingMessage> setOptOut(boolean optedOut) {
+        List<ReportingMessage> messageList = new LinkedList<>();
+        PilgrimSdk.stop(getContext());
+        messageList.add(new ReportingMessage(this, ReportingMessage.MessageType.OPT_OUT, System.currentTimeMillis(), null));
+        return messageList;
+    }
+
+    /**
+     * Returns ether last set user information, or a new
+     * PilgrimUserInfo instance.
+     */
+    private PilgrimUserInfo getUserInfo() {
+        PilgrimUserInfo info = PilgrimSdk.get().getUserInfo();
+        // can be null if it hasn't been set previously
+        if (info == null) {
+            info = new PilgrimUserInfo();
+        }
+        return info;
+    }
+
+    /* Section: User attribute set */
+    @Override
+    public void onSetAllUserAttributes(Map<String, String> attributes, Map<String, List<String>> attributeLists, FilteredMParticleUser filteredMParticleUser) {
+        // NOTE: attributeList not supported
+        // Is this ID correct?
+        updateUser(filteredMParticleUser.mpUser);
+        PilgrimUserInfo info = getUserInfo();
+        for (Map.Entry<String, String> entry : attributes.entrySet()) {
+            info.put(entry.getKey(), entry.getValue());
+        }
+        PilgrimSdk.get().setUserInfo(info);
+    }
+
+
+    @Override
+    public void onIncrementUserAttribute(String s, String s1, FilteredMParticleUser filteredMParticleUser) {
+        // Ignored,  Not supported atm
+    }
+
+    @Override
+    public void onRemoveUserAttribute(String s, FilteredMParticleUser filteredMParticleUser) {
+        PilgrimUserInfo info = getUserInfo();
+        // Do we want to log if it doesn't exist?
+        info.remove(s);
+    }
+
+    @Override
+    public void onSetUserAttribute(String s, Object o, FilteredMParticleUser filteredMParticleUser) {
+        PilgrimUserInfo info = getUserInfo();
+        info.put(s, o.toString());
+        // update
+        PilgrimSdk.get().setUserInfo(info);
+    }
+
+    @Override
+    public void onSetUserTag(String s, FilteredMParticleUser filteredMParticleUser) {
+        // Ignored, not supported atm.
+    }
+
+    @Override
+    public boolean supportsAttributeLists() {
+        // We don't support attribute lists
+        // the PilgrimSdk currently only
+        // allows single k => value attributes.
+        return false;
+    }
+
+    @Override
+    public void onSetUserAttributeList(String s, List<String> list, FilteredMParticleUser filteredMParticleUser) {
+        // Ignored, not supported
+    }
+
+    @Override
+    public void onConsentStateUpdated(ConsentState consentState, ConsentState consentState1, FilteredMParticleUser filteredMParticleUser) {
```

Comment:
how is consent state different than opt out?

Reply:
The consent is about GDPR and opt out is they removing the mParticle kit from their dashboard